src/Features/Core/Portable/Workspace/MiscellaneousFileUtilities.cs

```diff
@@ -0,0 +1,122 @@
+            var assemblyName = Guid.NewGuid().ToString("N");
+
+            var projectInfo = ProjectInfo.Create(
+                projectId,
+                VersionStamp.Create(),
+                name: FeaturesResources.Miscellaneous_Files,
+                assemblyName,
+                languageInformation.LanguageName,
+                compilationOptions: compilationOptions,
+                parseOptions: parseOptions,
+                documents: SpecializedCollections.SingletonEnumerable(documentInfo),
+                metadataReferences: metadataReferences);
+
+            // Miscellaneous files projects are never fully loaded since, by definition, it won't know
+            // what the full set of information is except when the file is script code.
+            return projectInfo.WithHasAllInformation(hasAllInformation: sourceCodeKind == SourceCodeKind.Script);
+        }
+
+        // Do not inline this to avoid loading Microsoft.CodeAnalysis.Scripting unless a script file is opened in the workspace.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static CompilationOptions GetCompilationOptionsWithScriptReferenceResolvers(HostWorkspaceServices services, CompilationOptions compilationOptions, string filePath)
+        {
+            var metadataService = services.GetRequiredService<IMetadataService>();
+
+            var baseDirectory = PathUtilities.GetDirectoryName(filePath);
+
+            // TODO (https://github.com/dotnet/roslyn/issues/5325, https://github.com/dotnet/roslyn/issues/13886):
+            // - Need to have a way to specify these somewhere in VS options.
+            // - Add default namespace imports, default metadata references to match csi.rsp
+            // - Add default script globals available in 'csi goo.csx' environment: CommandLineScriptGlobals
+
+            var referenceResolver = RuntimeMetadataReferenceResolver.CreateCurrentPlatformResolver(
+                searchPaths: ImmutableArray.Create(RuntimeEnvironment.GetRuntimeDirectory()),
+                baseDirectory: baseDirectory,
+                fileReferenceProvider: (path, properties) => metadataService.GetReference(path, properties));
+
+            return compilationOptions
+                .WithMetadataReferenceResolver(referenceResolver)
+                .WithSourceReferenceResolver(new SourceFileResolver(searchPaths: ImmutableArray<string>.Empty, baseDirectory));
+        }
+
+        private static SourceCodeKind GetSourceCodeKind(
+            ParseOptions? parseOptions,
+            string fileExtension,
+            LanguageInformation languageInformation)
+        {
+            if (parseOptions != null)
+            {
+                return parseOptions.Kind;
+            }
+
+            return string.Equals(fileExtension, languageInformation.ScriptExtension, StringComparison.OrdinalIgnoreCase) ?
+                SourceCodeKind.Script : SourceCodeKind.Regular;
+        }
+    }
+
+    internal class LanguageInformation
```

Comment:
Record?

Reply:
I get this error - CS0518	Predefined type 'System.Runtime.CompilerServices.IsExternalInit' is not defined or imported

Do we need to do something to enable records?